src/sources/util/tcp/request_limiter.rs

```diff
@@ -0,0 +1,158 @@
+use std::cmp::Ordering;
+use std::sync::{Arc, Mutex};
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+
+const EWMA_WEIGHT: f32 = 0.1;
```

Comment:
This seems low, but I don't know what is the variance of the values going into the EWMA and how fast you want this to respond to changes in the request size.

Reply:
In general I expect the values to have a fairly low variance. In most cases this is decided by the protocol used, but it's not something we can know for certain ahead of time, so I'm using EWMA to find the value.
There are 2 main cases.

The `socket` source, where each request generally has 1 event (or very few) events. The weight doesn't really matter here, each request is so small that the limiter will reach it's max (equal to the number of CPUs) quickly even with a low weight.

Sources where you can receive large requests, like `fluent`, where this limiter is more important. The average starts high so it is very limited at first since we don't know the true size of the requests yet, and it will eventually get an accurate average. In real-world tests, the size of the requests here are very stable (fluent buffers the requests up to a specific size).

Overall, I'm not really expecting the EWMA to adjust frequently, it's just to figure out the typical size of requests. I think a pretty large range of values for the EWMA_WEIGHT would be sufficient.